cpp/src/io/orc/writer_impl.cu

```diff
@@ -1937,6 +1944,87 @@ string_dictionaries allocate_dictionaries(orc_table_view const& orc_table,
           std::move(is_dict_enabled)};
 }
 
+struct string_length_functor {
+  __device__ inline size_type operator()(int const i) const
+  {
+    if (i >= num_chunks * 2) return 0;
+
+    auto const min = i % 2 == 0;
+    auto const idx = i / 2;
+    auto& str_val =
+      min ? stripe_stat_chunks[idx].min_value.str_val : stripe_stat_chunks[idx].max_value.str_val;
+    auto const str = stripe_stat_merge[idx].stats_dtype == dtype_string;
+    return str ? str_val.length : 0;
+  }
+
+  int num_chunks;
+  statistics_chunk* stripe_stat_chunks;
+  statistics_merge_group* stripe_stat_merge;
+};
+
+__global__ void copy_string_data(char* string_pool,
+                                 size_type* offsets,
+                                 statistics_chunk* chunks,
+                                 statistics_merge_group const* groups)
+{
+  auto const idx = blockIdx.x / 2;
+  if (groups[idx].stats_dtype == dtype_string) {
```

Comment:
Unfortunately, we don't have a benchmark for the ORC chunked writer. Should be quick to implement one based on the Parquet chunked writer benchmark.
The data gen utilities also allows to tune the data profile and check if there is a case where we get a big perf hit.

Reply:
I worried about this as well. I don't know if we can have a "typical" dataset here. I don't like marching all the columns and only some blocks have work. I would imagine the number of string columns could dominate a table or be a very small minority. I would be happy to investigate building a mask, but I worry about the general performance. Since we are dealing with min and max strings, that means that some blocks have huge things to copy and some have tiny things. In the end, I decided to get this out there working and see if we wanted to pursue anything more complex. Any input on this, @vuule ?

Reply:
I think setting up a benchmark and covering different cases should be fairly quick, since we have all scaffolding. Changing the column types, null frequency, avg string length are all trivial with the data generator we have.
I would suggest to implement this benchmark and use it to eval if there are truly terrible cases due to the changes here. 
We need ORC chunked writer benchmark anyway 😁

Reply:
I can handle that, I'll add it and report some results.

Reply:
I don't know if my calculation for bytes copied is correct, but there are the times:

new code:
```
--------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                      Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------------------------------------------------
OrcWrite/3Gb8Cols/8/manual_time                                              191 ms          190 ms            3 bytes_per_second=2.61845G/s encoded_file_size=225.369M peak_memory_usage=776.914M
OrcWrite/3Gb1024Cols/1024/manual_time                                       1385 ms         1385 ms            1 bytes_per_second=369.627M/s encoded_file_size=119.05M peak_memory_usage=7.1454G
OrcWriteChunked/3Gb8Cols64Chunks/8/8/iterations:4/manual_time                434 ms          434 ms            4 bytes_per_second=1.15159G/s encoded_file_size=302.516M peak_memory_usage=102.262M
OrcWriteChunked/3Gb1024Cols64Chunks/1024/8/iterations:4/manual_time         1995 ms         1995 ms            4 bytes_per_second=256.595M/s encoded_file_size=460.356M peak_memory_usage=894.068M
OrcWriteChunked/3Gb8Cols128Chunks/8/64/iterations:4/manual_time             1301 ms         1301 ms            4 bytes_per_second=393.466M/s encoded_file_size=315.134M peak_memory_usage=16.9397M
OrcWriteChunked/3Gb1024Cols128Chunks/1024/64/iterations:4/manual_time       6743 ms         6743 ms            4 bytes_per_second=75.9305M/s encoded_file_size=592.529M peak_memory_usage=832.343M
```
branch-22.06:
```
--------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                      Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------------------------------------------------
OrcWrite/3Gb8Cols/8/manual_time                                              190 ms          190 ms            3 bytes_per_second=2.62546G/s encoded_file_size=225.369M peak_memory_usage=776.914M
OrcWrite/3Gb1024Cols/1024/manual_time                                       1381 ms         1381 ms            1 bytes_per_second=370.782M/s encoded_file_size=119.05M peak_memory_usage=7.1454G
OrcWriteChunked/3Gb8Cols64Chunks/8/8/iterations:4/manual_time                432 ms          432 ms            4 bytes_per_second=1.15767G/s encoded_file_size=302.515M peak_memory_usage=102.262M
OrcWriteChunked/3Gb1024Cols64Chunks/1024/8/iterations:4/manual_time         1966 ms         1966 ms            4 bytes_per_second=260.418M/s encoded_file_size=460.265M peak_memory_usage=894.039M
OrcWriteChunked/3Gb8Cols128Chunks/8/64/iterations:4/manual_time             1300 ms         1300 ms            4 bytes_per_second=393.941M/s encoded_file_size=315.134M peak_memory_usage=16.9371M
OrcWriteChunked/3Gb1024Cols128Chunks/1024/64/iterations:4/manual_time       6430 ms         6430 ms            4 bytes_per_second=79.6221M/s encoded_file_size=592.432M peak_memory_usage=831.99M
```

Reply:
I think if anything what I have found through testing is that the statistics writing portion of the code is a tiny fraction of the total runtime and the amount of statistics data is simply not large. I'm unsure if this helps or clouds the matter though.

Reply:
I think that helps tell us that we shouldn't waste many cycles optimizing this code if it's not a major bottleneck. I'm on board with getting this merged as is and then doing a deeper dive if you have time afterwards to see if this kernel ends up being latency- instead of compute-bound because of low GPU utilization from this conditional.